doc/threadsafe_function.md

```diff
@@ -0,0 +1,303 @@
+# ThreadSafeFunction
+
+JavaScript functions can normally only be called from a native addon's main
+thread. If an addon creates additional threads, then node-addon-api functions
+that require a `Napi::Env`, `Napi::Value`, or `Napi::Reference` must not be
+called from those threads.
+
+When an addon has additional threads and JavaScript functions need to be invoked
+based on the processing completed by those threads, those threads must
+communicate with the addon's main thread so that the main thread can invoke the
+JavaScript function on their behalf. The thread-safe function APIs provide an
+easy way to do this.
+
+These APIs provide the type `Napi::ThreadSafeFunction` as well as APIs to
+create, destroy, and call objects of this type.
+`Napi::ThreadSafeFunction::New()` creates a persistent reference that holds a
+JavaScript function which can be called from multiple threads. The calls happen
+asynchronously. This means that values with which the JavaScript callback is to
+be called will be placed in a queue, and, for each value in the queue, a call
+will eventually be made to the JavaScript function.
+
+`Napi::ThreadSafeFunction` objects are destroyed when every thread which uses
+the object has called `Release()` or has received a return status of
+`napi_closing` in response to a call to `BlockingCall()` or `NonBlockingCall()`.
+The queue is emptied before the `Napi::ThreadSafeFunction` is destroyed. It is
+important that `Release()` be the last API call made in conjunction with a given
```

Comment:
One thing (even with the current wording, generally taken from [n_api_asynchronous_thread_safe_function_calls](https://nodejs.org/api/n-api.html#n_api_asynchronous_thread_safe_function_calls)): I don't think the text is correct, because `Release()` [or `Abort()`] don't _necessarily_ have to be the last call. 

In a normal, non-aborted flow, `Release()` would be the last call, but in an aborted flow, it is possible that a `[Non]BlockingCall()` could be the last call, as it also decreases the thread count:

>  This can be used as a criterion for terminating the thread.

Should we better specify the abortable behavior here as well, or is it fine to keep in the Abort section?

Reply:
`Release()` or `Abort()` *absolutely* have to be the last call – on the thread making the call. I think the most important thing to point out, and the aspect to emphasize here is this: After you've called `Release()` or `Abort()`, make no further use of the TSFN *on that thread*, irrespective of how this will affect the other threads – namely that, in the case of `Abort()` it will cause them to receive `napi_closing`s.

This latter point is documented further down in the paragraph, and `BlockingCall()` and `NonBlockingCall()` are documented to have this feature such that the two calls are presented side-by-side as equal and alternative ways of being informed of the TSFN's impending closure.

The fact is that the documentation in core is almost certainly not phrased ideally. This is actually an opportunity to improve both places of documentation.